File: selection/selection_parallel.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits>
#include <thread>
#include <mutex>
#include <future>

using namespace std;

void selectionSortParallel(vector<int> &arr) {
    int n = arr.size();

    for (int i = 0; i < n - 1; ++i) {
        int min_idx = i;
        
        // Use multiple threads to find minimum in parallel
        int num_threads = min(4, (int)thread::hardware_concurrency());
        if (num_threads == 0) num_threads = 2;
        
        vector<future<int>> futures;
        int chunk_size = (n - i - 1) / num_threads;
        
        for (int t = 0; t < num_threads; ++t) {
            int start = i + 1 + t * chunk_size;
            int end = (t == num_threads - 1) ? n : i + 1 + (t + 1) * chunk_size;
            
            futures.push_back(async(launch::async, [&arr, start, end, i]() {
                int local_min_idx = i;
                for (int j = start; j < end; ++j) {
                    if (arr[j] < arr[local_min_idx]) {
                        local_min_idx = j;
                    }
                }
                return local_min_idx;
            }));
        }
        
        // Collect results and find global minimum
        for (auto& future : futures) {
            int local_min = future.get();
            if (arr[local_min] < arr[min_idx]) {
                min_idx = local_min;
            }
        }

        swap(arr[i], arr[min_idx]);
    }s
}

void printArray(const vector<int> &arr) {
    for (int val : arr) {
        cout << val << " ";
    }
    cout << endl;
}

int main() {
    vector<int> arr = {64, 25, 12, 22, 11};

    cout << "Original array: ";
    printArray(arr);

    selectionSortParallel(arr);

    cout << "Sorted array: ";
    printArray(arr);

    return 0;
}

File: selection/selection_parallel.py
# parallelizes inner loop of selection sort

from concurrent.futures import ProcessPoolExecutor

def find_min_index(arr_slice, offset):
    """
    Find the index of the minimum element in the slice,
    returning the index relative to the full array.
    """
    min_idx = 0
    for i in range(1, len(arr_slice)):
        if arr_slice[i] < arr_slice[min_idx]:
            min_idx = i
    return min_idx + offset

def selection_sort_parallel(arr):
    n = len(arr)
    for i in range(n - 1):
        # Parallel part: find the min index from i to n-1
        with ProcessPoolExecutor() as executor:
            # Only split if large enough to benefit
            if n - i > 1000:
                # Divide the remaining unsorted part into chunks
                num_chunks = 4
                chunk_size = (n - i) // num_chunks
                futures = []
                for c in range(num_chunks):
                    start = i + c * chunk_size
                    end = i + (c + 1) * chunk_size if c != num_chunks - 1 else n
                    futures.append(executor.submit(find_min_index, arr[start:end], start))

                # Get min index from among the chunks
                min_idx = i
                for future in futures:
                    idx = future.result()
                    if arr[idx] < arr[min_idx]:
                        min_idx = idx
            else:
                # Fallback to sequential for small segments
                min_idx = i
                for j in range(i + 1, n):
                    if arr[j] < arr[min_idx]:
                        min_idx = j

        arr[i], arr[min_idx] = arr[min_idx], arr[i]

def print_array(arr):
    for val in arr:
        print(val, end=" ")
    print()

if __name__ == "__main__":
    arr = [64, 25, 12, 22, 11, 100, 4000, 2400]
    
    print("Original array: ", end="")
    print_array(arr)
    
    selection_sort_parallel(arr)
    
    print("Sorted array: ", end="")
    print_array(arr)
